Devise Authentication for API

kianaditya opened the pull request:
# PT Link

[Devise Authentication](https://www.pivotaltracker.com/story/show/163348231)

# User Story

```
As a developer
In order to restrict access
I would like to require login for the user to see articles on my mobile app
```
# Screenshot

![image](https://user-images.githubusercontent.com/25456343/51674537-3176b580-1fd1-11e9-9568-e747b1cc05c0.png)


# Description of PR

Uses devise token auth gem to set up token based user authentication

kianaditya committed:
updates credentials

kianaditya committed:
Fixes merge conflicts

kianaditya committed:
fixes credentials

kianaditya committed:
Merge branch 'develop' of github.com:CraftAcademy/CA_Newsroom_team_1 into develop

kianaditya committed:
Merge branch 'develop' of github.com:CraftAcademy/CA_Newsroom_team_1 into develop

kianaditya committed:
Updates schema

kianaditya committed:
Updates seeds file
Adds specs from registration and session
Adds devise token auth gem

kianaditya committed:
Adds controllers for registration and session

kianaditya committed:
Updaets user model,
updates api controller
creates migration to connect devise token auth with users

tochman commented on spec/requests/api/registrations_spec.rb:
```diff
@@ -0,0 +1,61 @@
+    end
+
+    it 'an already registered email' do
+      FactoryBot.create(:user, email: 'me@mail.com',
```
Please omit the module name

kianaditya replied:
Fixed

tochman commented on spec/requests/api/registrations_spec.rb:
```diff
@@ -0,0 +1,61 @@
+                                 first_name: 'John',
+                                 last_name: 'Doe',
+                              }, headers: headers
+
```
I would extract JSON.parse(response.body) to a helper method. A bit WET to repeat that in every test.

kianaditya replied:
Fixed, thanks for guidance

tochman commented on spec/requests/api/registrations_spec.rb:
```diff
@@ -0,0 +1,61 @@
+                              }, headers: headers
+
+      expect(JSON.parse(response.body)['errors']['email']).to eq ['is not an email']
+      expect(response.status).to eq 422
```
I know that O has two expect statements in his examples, but I would like you to have one expect per it block. Okay?

kianaditya replied:
Updated as per my understanding. Please have a look

tochman commented on app/models/user.rb:
```diff
@@ -3,6 +3,7 @@ class User < ApplicationRecord
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
+         include DeviseTokenAuth::Concerns::User
```
Indent this

kianaditya replied:
fixed

tochman commented on config/credentials.yml.enc:
```diff
@@ -1 +1 @@
-WNp7kxC6TQuzj7mT5dk2D1Lq1P09OdUloZgrvU3+M8mH0M5vRuzCxAOExpXhx6uDnJj0ujb2TUZAbBTAX6wahb43xf4ifZnXrvXPwrUhjRU6Mlzd8q4Mfm7g+H4rJjJ+auXtxvc6p2+JgKnWMlh4VlLJRJcY384An6GnF21m7V8qAgolDZBVBVhsEQTVcSY4WOcGE2tRM5HiGudyuZbUagxDg6K9JHda8fB+F8C63YC8UMqdHdE8KNj4YwytoeA7lM6DiL3YLFv5u9cpSGhiv6hX2r4BfAnQivP4eilDzn55R/kS86tx4NNBspUC5e2kQo8kJY49bOBIcZQ0UJn7HtY9FWHgy1UH8aJLacWkr324iQgZR47fGWykYLmW3tX7cYqmiDlQ0QZUL6O ...
\ No newline at end of file
+ZNfHNHDXAg+RPLswfcUFQwG+rNnPo5OHx5KbKylBcN4gPY2voXKy/ECJXkiVkFKhau+lIFFwoVjBdJs9X4mfzyjG0C2t9U1BvEKvI51Eib3npkL/J7tzFvKkJAXdxQYbYui1qQgEbCv20oYgh9tJ/YIU3Tk3q4niOlOdXAsbtk/lJIqVih76XGg0enLj7yGGSaO6wLqJ2nCEo3M4kxSg7xy1EjglAgUmxFgP54BR6N0uxImNdac1FNYPdIKnKsErYl8kw2AwQlgdIIkv/niZX0SOLZfUekHP5j4HtKhIjPvlHH41H+cDtP0MbIyjcI2LAPp/PeyV/ahfu4MlNG52S0vajlvdv6bmD1i1Odcnv7xLwIfXLvK+IA3WBz1yj4GthR+1SVxkD3niX3R ...
```
Did you add any new credentials in this PR?

kianaditya replied:
No new credentials added. I had to update credentials because some of the tests were failing for me but not for others.

tochman commented on spec/requests/api/sessions_spec.rb:
```diff
@@ -0,0 +1,42 @@
+require 'rails_helper'
+
+RSpec.describe 'Sessions', type: :request do
+  let(:user) { FactoryBot.create(:user) }
```
Same thing here. You can omit FactoryBot.

kianaditya replied:
Fixed

kianaditya committed:
removes factorybot module from spec tests

kianaditya committed:
minor formatting fix

kianaditya committed:
refactors registrations spec file

kianaditya committed:
Refactor sessions_spec

kianaditya committed:
Updates applications controller to skip auth token check,
deletes unrequired controllers,
extracts json reponse in a helper method and refactors spec files accordingly

kianaditya committed:
Updates cors setting in application.rb

tochman reviewed (approved):
All good
